Sources/NIOHTTP1/HTTPTypes.swift

```diff
@@ -176,33 +176,332 @@ struct HTTPHeader {
+    public static let xRequestedWith = HTTPHeaderName("X-Requested-With")
+    public static let xFrameOptions = HTTPHeaderName("X-Frame-Options")
+}
+
+/// Different types representing HTTP header names.
+private enum HTTPHeaderNameType {
+    /// Header name represented by a `String`.
+    case string(String)
+    /// Header name represented by a `HTTPHeaderName`.
+    case headerName(HTTPHeaderName)
+
+    /// Returns a lowercased version of the header name.
+    func lowercased() -> String {
+        switch self {
+        case .headerName(let name):
+            return name.lowerCaseName
+        case .string(let string):
+            return string.lowercased()
+        }
+    }
+}
+
+/// Extensions dealing with `HTTPHeaderNameType`s.
 private extension ByteBuffer {
-    func equalCaseInsensitiveASCII(view: String.UTF8View, at index: HTTPHeaderIndex) -> Bool {
-        guard view.count == index.length else {
+    /// Writes the `HTTPHeaderNameType` and also the seperator (`: `). It returns the length of the name (without the seperator).
+    mutating func writeWithSeperator(nameType: HTTPHeaderNameType) -> Int {
+        switch nameType {
+        case .headerName(let name):
+            return self.write(bytes: name.bytesWithSeperator) - 2
+        case .string(let s):
+            let len = self.write(string: s)!
+            self.write(staticString: headerSeparator)
+            return len
+        }
+    }
+
+    func equalCaseInsensitiveASCII(nameType: HTTPHeaderNameType, at index: HTTPHeaderIndex) -> Bool {
+        switch nameType {
+        case .headerName(let name):
+            return self.equalCaseInsensitiveASCII(name: name, at: index)
+        case .string(let string):
+            return self.equalCaseInsensitiveASCII(view: string.utf8, at: index)
+        }
+    }
+
+    private func equalCaseInsensitiveASCII(name: HTTPHeaderName, at index: HTTPHeaderIndex) -> Bool {
+        guard name.count == index.length else {
             return false
         }
         return withVeryUnsafeBytes { buffer in
             // This should never happens as we control when this is called. Adding an assert to ensure this.
             assert(index.start <= self.capacity - index.length)
+
             let address = buffer.baseAddress!.assumingMemoryBound(to: UInt8.self)
-            for (idx, byte) in view.enumerated() {
-                guard byte.isASCII && address.advanced(by: index.start + idx).pointee & 0xdf == byte & 0xdf else {
+            for (idx, byte) in name.lowerCaseBytes.enumerated() {
+                guard address.advanced(by: index.start + idx).pointee.maskedLowercase == byte else {
                     return false
                 }
             }
             return true
         }
     }
-}
 
-
-private extension UInt8 {
-    var isASCII: Bool {
-        return self <= 127
+    private func equalCaseInsensitiveASCII(view: String.UTF8View, at index: HTTPHeaderIndex) -> Bool {
```

Comment:
both those `equalCaseInsensitiveASCII` should be expressible as functions on `<S: Sequence>(bytes: S) where S.Element == UInt8`

Reply:
I think this lacks .count

Reply:
@normanmaurer then on `Collection`